### PR TITLE
Separate validation from metadata retrieval

### LINKED ISSUE #496: Validate Schema separately to getMetaData

### PR DESCRIPTION
**Pull request by muffinresearch**
Fixes #496

### CODE REVIEW

**Comment by kumar303** [comment]
I would be ok with this but a method that depends on another is always a sign of clunkiness to me. The question to ask is: would you ever need to get metadata without validating the manifest first? I think the answer is no. If that's true, you could change the class to this:
```` 
export default class ManifestJSONParser {

  constructor(jsonString, collector) {
    this.parsedJSON = JSON.parse(jsonString);
    this.collector = collector;
    this.isValid = this._validate();
  }
  ...
}
````
The only downside is it conceals the fact that some potentially costly validation is taking place inside the constructor. I think the upside of making the `getMetaData()` function more easy to work with is worth the trade-off though.

r+wc

**Comment by muffinresearch** [comment]
It's all swings and roundabouts!

I was working on the basis that the constructor shouldn't do too much, but I see what you're saying and I think you're right -  it's probably a touch cleaner your way on balance.

**Comment by muffinresearch** [comment]
@kumar303 ok this is ready for another look.

File: src/linter.js [hunk 1]
```diff
@@ -223,9 +223,8 @@ export default class Linter {
           log.info('Retrieving metadata from manifest.json');
           return this.io.getFileAsString(MANIFEST_JSON)
             .then((json) => {
-              var result = new ManifestJSONParser(
-                json, this.collector).getMetadata();
-              return result.metadata;
+              var manifestParser = new ManifestJSONParser(json, this.collector);
```
Comment (kumar303) [inline src/linter.js]: could be `let manifestParser...`

File: tests/parsers/test.manifestjson.js [hunk 2]
```diff
@@ -13,8 +13,8 @@ describe('ManifestJSONParser manifestVersion', function() {
     var json = validManifestJSON({manifest_version: 'whatever'});
     var manifestJSONParser = new ManifestJSONParser(json,
                                                     addonLinter.collector);
-    var result = manifestJSONParser.getMetadata();
-    assert.equal(result.isValid, false);
+    assert.equal(manifestJSONParser.isValid, false);
+    manifestJSONParser.getMetadata();
```
Comment (kumar303) [inline tests/parsers/test.manifestjson.js]: does this have some side effect? Why does it need to be called here?

File: tests/parsers/test.manifestjson.js [hunk 3]
```diff
@@ -26,8 +26,9 @@ describe('ManifestJSONParser manifestVersion', function() {
     var json = validManifestJSON({manifest_version: '1'});
     var manifestJSONParser = new ManifestJSONParser(json,
                                                     addonLinter.collector);
-    var result = manifestJSONParser.getMetadata();
-    assert.equal(result.isValid, false);
+    assert.equal(manifestJSONParser.isValid, false);
+
+    manifestJSONParser.getMetadata();
```
Comment (kumar303) [inline tests/parsers/test.manifestjson.js]: I can't spot any side effects that would happen from calling this so it looks unnecessary.

**Comment by kumar303** [comment]
r+wc

File: tests/parsers/test.manifestjson.js (hunk 2 above)
Comment (muffinresearch) [inline tests/parsers/test.manifestjson.js]: Oh duh - that's a leftover - well spotted.

File: src/linter.js (hunk 1 above)
Comment (muffinresearch) [inline src/linter.js]: I'm tending to use `let` when needed. Inside a function like this there's no difference between `var` and `let`.

Comment (muffinresearch) [inline src/linter.js]: ...and even then I forget to sometimes - old habits die hard.

Comment (kumar303) [inline src/linter.js]: Whatever works. I've been trying to always use let unless var is needed. My rationale is that a code refactor might introduce a new block (a line down from your closure, for example) that could leak the variable.

**Pull request merged**